### PR TITLE
[BEAM-11205] Libraries-BOM 20.0.0 with gRPC and protobuf

### PR DESCRIPTION
**Pull request by suztomo**
Upgrading the libraries-bom version to 20.0.0 along with gRPC and protobuf.

# Changes in mock object setup in failed tests

Some GCP-related tests failed. The latest google-http-client internally calls `getStatusCode()` twice per HTTP response. This change broke some tests where their mock setups had assumed that the method call occurs only once per request (and the tests were reusing the same mock response object for different responses). I updated the mock setup so that it now prepares different mock response objects for different responses.

# Linkage Check

No new linkage errors. https://gist.github.com/suztomo/ac11c74fb24ebb8ac0e11791a8be11a9

Initially it had problem with old gax-httpjson used by google-cloud-bigquerystorage ([log](https://gist.github.com/suztomo/dda7011d4dd8471289b0a7847e9afaf4#file-gistfile1-txt-L962)). Declaring gax-httpjson as a dependency of google-cloud-platform module fixed the linkage errors.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>SDK</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by suztomo** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by suztomo** [comment]
Run PythonLint PreCommit

**Comment by suztomo** [comment]
Run PythonLint PreCommit

**Comment by suztomo** [comment]
`:sdks:java:core:compileTestJava` failed:

```
/home/jenkins/jenkins-slave/workspace/beam_PreCommit_SQL_Commit/src/sdks/java/core/src/test/java/org/apache/beam/sdk/metrics/MetricsTest.java:450: error: incompatible types: inferred type does not conform to lower bound(s) |  
-- | --
  | assertThat( |  
  | ^ |  
  | inferred: MetricResult<DistributionResult> |  
  | lower bound(s): CAP#1,MetricResult<DistributionResult> |  
  | where CAP#1 is a fresh type-variable: |  
  | CAP#1 extends Object super: MetricResult<DistributionResult> from capture of ? super MetricResult<DistributionResult> |  
  | Note: Some input files use or override a deprecated API. |  
  | Note: Recompile with -Xlint:deprecation for details. |  
  | Note: Some input files use unchecked or unsafe operations. |  
  | Note: Recompile with -Xlint:unchecked for details. |  
  | 1 error
```

**Comment by suztomo** [comment]
Other errors

```
java.lang.AssertionError: 
Expected: an instance of java.io.FileNotFoundException
     but: <java.io.IOException: Error executing batch GCS request> is a java.io.IOException
Stacktrace was: java.io.IOException: Error executing batch GCS request
	at org.apache.beam.sdk.extensions.gcp.util.GcsUtil.executeBatches(GcsUtil.java:585)
	at org.apache.beam.sdk.extensions.gcp.util.GcsUtil.getObjects(GcsUtil.java:314)
	at org.apache.beam.sdk.extensions.gcp.util.GcsUtil.fileSizes(GcsUtil.java:363)
	at org.apache.beam.sdk.extensions.gcp.util.GcsUtilTest.testGetSizeBytesWhenFileNotFoundBatchRetry(GcsUtilTest.java:523)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExpectedException$ExpectedExceptionStatement.evaluate(ExpectedException.java:258)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:119)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.util.concurrent.ExecutionException: java.lang.NullPointerException
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1908)
	at org.apache.beam.sdk.util.MoreFutures.get(MoreFutures.java:60)
	at org.apache.beam.sdk.extensions.gcp.util.GcsUtil.executeBatches(GcsUtil.java:577)
	... 54 more
Caused by: java.lang.NullPointerException
	at com.google.api.client.googleapis.batch.BatchUnparsedResponse.trimCrlf(BatchUnparsedResponse.java:276)
	at com.google.api.client.googleapis.batch.BatchUnparsedResponse.readLine(BatchUnparsedResponse.java:272)
	at com.google.api.client.googleapis.batch.BatchUnparsedResponse.parseNextResponse(BatchUnparsedResponse.java:97)
	at com.google.api.client.googleapis.batch.BatchRequest.execute(BatchRequest.java:267)
	at org.apache.beam.sdk.extensions.gcp.util.GcsUtil.lambda$executeBatches$0(GcsUtil.java:573)
	at org.apache.beam.sdk.util.MoreFutures.lambda$runAsync$2(MoreFutures.java:140)
	at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1640)
```

Screen Shot 2021-04-14 at 10 43 55

trimCrlf is getting `null`. This implies the mock response doesn't have enough values.

Screen Shot 2021-04-14 at 10 49 21

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImplTest.java [hunk 1]
```diff
@@ -1056,11 +1082,25 @@ public void testCreateTableRetry() throws IOException {
...
+    verify(response, times(1)).getContent();
+    verify(response, times(1)).getContentType();
+
+    // This was null
+    // verifyNotNull(ret.getTableReference());
+
+    // Google-http-client 1.39.1 does not read the content of response with error status code. How
+    // can we ensure our library shows proper error messages to users?
     expectedLogs.verifyInfo(
         "Quota limit reached when creating table project:dataset.table, "
```
Comment (suztomo) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImplTest.java:1105]: Before this PR, `ApiErrorExtractor` was reading the GoogleJsonResponseException that holds the content:

```
{
  "code" : 403,
  "errors" : [ {
    "domain" : "global",
    "reason" : "rateLimitExceeded"
  } ],
  "message" : "rateLimitExceeded"
}
```

In tryCreateTable ([code](https://github.com/apache/beam/blob/4b7b74673b647c8d964b4877a8d66d47096acce4/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImpl.java#L626)), `extractor.rateLimited(e)` determines the cause of the problem.

File: sdks/java/extensions/google-cloud-platform-core/src/test/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtilTest.java [hunk 2]
```diff
@@ -502,7 +502,9 @@ public void testGetSizeBytesWhenFileNotFoundBatchRetry() throws Exception {
 
     // 429: Too many requests, then 200: OK.
     when(mockResponse.getStatusCode()).thenReturn(429, 200);
-    when(mockResponse.getContent()).thenReturn(toStream("error"), toStream(content));
+    // Google-http-client:1.39.1 and above does not read content stream of the error response.
```
Comment (elharo) [inline sdks/java/extensions/google-cloud-platform-core/src/test/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtilTest.java:505]: This should not be true unless disconnect is being called somewhere.

Comment (suztomo) [reply sdks/java/extensions/google-cloud-platform-core/src/test/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtilTest.java:505]: Thanks. I'll create a test case for google-http-java-client to verify this.

Comment (suztomo) [reply sdks/java/extensions/google-cloud-platform-core/src/test/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtilTest.java:511]: It turned out that it was mock setup that had wrong assumption that getStatusCode is called once per response.

**Comment by suztomo** [comment]
Run Java Postcommit

**Comment by suztomo** [comment]
Run GoPortable PreCommit

**Comment by suztomo** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by suztomo** [comment]
Run Python PreCommit

**Comment by suztomo** [comment]
Run PostCommit_Java_Hadoop_Versions

**Comment by suztomo** [comment]
Run PostCommit_Java_Dataflow

**Comment by suztomo** [comment]
Run PostCommit_Java_DataflowV2

**Comment by suztomo** [comment]
Run Java HadoopFormatIO Performance Test

**Comment by suztomo** [comment]
Run BigQueryIO Streaming Performance Test Java

**Comment by suztomo** [comment]
Run Dataflow ValidatesRunner

**Comment by suztomo** [comment]
Run Spark ValidatesRunner

**Comment by suztomo** [comment]
Run SQL Postcommit

**Comment by suztomo** [comment]
Run Java PreCommit

**Comment by suztomo** [comment]
Run Java HadoopFormatIO Performance Test

**Comment by suztomo** [comment]
Run Java HadoopFormatIO Performance Test

**Comment by suztomo** [comment]
Run Java PreCommit

**Comment by suztomo** [comment]
Run Java PreCommit

**Comment by suztomo** [comment]
It creates Linkage Errors due to the old google-cloud-bigquerystorage artifact

https://gist.github.com/suztomo/dda7011d4dd8471289b0a7847e9afaf4

**Comment by suztomo** [comment]
Run Dataflow ValidatesRunner

**Comment by suztomo** [comment]
Run Spark ValidatesRunner

**Comment by suztomo** [comment]
Run SQL Postcommit

**Comment by suztomo** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by suztomo** [comment]
Problem in ':grpc-all:dependencies' task over JUnit dependencies.

```
Execution failed for task ':grpc-all:dependencies'.
> Could not resolve all dependencies for configuration ':grpc-all:runtimeClasspath'.
   > Conflict(s) found for the following module(s):
       - junit:junit between versions 4.13.1 and 4.12
     Run with:
         --scan or
         :grpc-all:dependencyInsight --configuration runtimeClasspath --dependency junit:junit
     to get more insight on how to solve the conflict.
```

File: sdks/java/extensions/google-cloud-platform-core/src/test/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtilTest.java [hunk 3]
```diff
@@ -497,21 +497,30 @@ public void testGetSizeBytesWhenFileNotFoundBatchRetry() throws Exception {
             + "\n";
     thrown.expect(FileNotFoundException.class);
 
-    final LowLevelHttpResponse mockResponse = Mockito.mock(LowLevelHttpResponse.class);
-    when(mockResponse.getContentType()).thenReturn("multipart/mixed; boundary=" + contentBoundary);
+    final LowLevelHttpResponse[] mockResponses =
+        new LowLevelHttpResponse[] {
+          Mockito.mock(LowLevelHttpResponse.class), Mockito.mock(LowLevelHttpResponse.class),
```
Comment (suztomo) [inline sdks/java/extensions/google-cloud-platform-core/src/test/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtilTest.java:502]: 2 mock objects for 2 responses.

File: sdks/java/extensions/google-cloud-platform-core/src/test/java/org/apache/beam/sdk/extensions/gcp/util/RetryHttpRequestInitializerTest.java [hunk 4]
```diff
@@ -209,23 +231,30 @@ public void testThrowIOException() throws IOException {
...
               int n = 0;
 
               @Override
-              public Integer answer(InvocationOnMock invocation) {
-                return n++ < retries ? 503 : 9999;
+              public MockLowLevelHttpResponse answer(InvocationOnMock invocation) throws Throwable {
+                MockLowLevelHttpResponse response = mock(MockLowLevelHttpResponse.class);
+                responses.add(response);
+                when(response.getStatusCode()).thenReturn(n++ < retries ? 503 : 9999);
+                return response;
```
Comment (suztomo) [inline sdks/java/extensions/google-cloud-platform-core/src/test/java/org/apache/beam/sdk/extensions/gcp/util/RetryHttpRequestInitializerTest.java:257]: Previously it was reusing the same mock object for different responses. Now it prepares different mock objects for different responses.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImplTest.java [hunk 5]
```diff
@@ -89,28 +91,35 @@
...
 import org.mockito.MockitoAnnotations;
 
 /** Tests for {@link BigQueryServicesImpl}. */
 @RunWith(JUnit4.class)
 public class BigQueryServicesImplTest {
   @Rule public ExpectedException thrown = ExpectedException.none();
   @Rule public ExpectedLogs expectedLogs = ExpectedLogs.none(BigQueryServicesImpl.class);
-  @Mock private LowLevelHttpResponse response;
+  // A test can make mock responses through setupMockResponses
+  private LowLevelHttpResponse[] responses;
```
Comment (suztomo) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImplTest.java:102]: Previously one mock response object was reused throughout multiple HTTP responses. This does not work any more as newer google-http-client calls `getStatusCode()` multiple times per response.

Now this test prepares different mock response objects for different responses.

**Comment by suztomo** [comment]
Run Dataflow ValidatesRunner

**Comment by suztomo** [comment]
Run PostCommit_Java_Dataflow

**Comment by suztomo** [comment]
Run PostCommit_Java_DataflowV2

**Comment by suztomo** [comment]
Run Java Postcommit

**Comment by suztomo** [comment]
R: @kennknowles @aaltay

**Comment by suztomo** [comment]
Run PostCommit_Java_DataflowV2

**Comment by suztomo** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by suztomo** [comment]
Run PostCommit_Java_DataflowV2

**Comment by suztomo** [comment]
34 successful checks!

**Comment by suztomo** [comment]
Run Java Precommit

**Review by Fokko — APPROVED** [review]
Thanks for picking this up @suztomo Looking good!

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImplTest.java (hunk 1 above)
Comment (suztomo) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImplTest.java:1230]: It was wrong mock object setup which expected getStatusCode was called only once per response.

**Comment by suztomo** [comment]
Run PostCommit_Java_DataflowV2

**Comment by suztomo** [comment]
Run Java PostCommit

**Comment by suztomo** [comment]
Run PostCommit_Java_Dataflow

**Comment by suztomo** [comment]
Run Dataflow ValidatesRunner

**Comment by suztomo** [comment]
Run SQL Postcommit

File: sdks/java/extensions/google-cloud-platform-core/src/test/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtilTest.java (hunk 3 above)
Comment (suztomo) [reply sdks/java/extensions/google-cloud-platform-core/src/test/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtilTest.java:502]: And today I came to know @kennknowles knew this 18 days ago already.

https://github.com/apache/beam/pull/14218#issuecomment-812091162

**Comment by suztomo** [comment]
R: @pabloem 
Can you review this PR?  It's almost same as https://github.com/apache/beam/pull/14218 but the mock object setup fix included.

**Comment by suztomo** [comment]
Run PythonDocker PreCommit

**Comment by suztomo** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by suztomo** [comment]
Run Portable_Python PreCommit

**Comment by suztomo** [comment]
Run PythonDocker PreCommit

**Comment by pabloem** [comment]
Run PostCommit_Java_Dataflow

**Review by pabloem — APPROVED** [review]
Thanks @suztomo ! LGTM
Let's just make sure that our Java tests are all passing, but feel free to merge after that.

**Comment by suztomo** [comment]
@pabloem Thank you. All Java tests passed. (Portable_Python and PythonDocker have been failing in pull requests in this repo)

**Pull request merged**